File: uhrwerk-config/src/main/java/io/qimia/uhrwerk/config/representation/Source.java
package io.qimia.uhrwerk.config.representation;


import io.qimia.uhrwerk.config.ConfigException;

import java.util.Arrays;

public class Source{

    private String connection_name;
    private String path;
    private String format;
    private String version;
    private Partition partition;
    private ParallelLoad parallel_load;
    private Select select;
    private Boolean autoloading = true;

    public Source() {}

    public Boolean getAutoloading() {
        return autoloading;
    }

    public void setAutoloading(Boolean autoloading) {
        this.autoloading = autoloading;
    }

    public String getConnection_name() {
        return connection_name;
    }

    public void setConnection_name(String connection_name) {
        this.connection_name = connection_name;
    }

    public String getPath() {
        return path;
    }

    public void setPath(String path) {
        this.path = path;
    }

    public String getFormat() {
        return format;
    }

    public void setFormat(String format) {
        this.format = format;
    }

    public String getVersion() {
        return version;
    }

    public void setVersion(String version) {
        this.version = version;
    }

    public Partition getPartition() {
        return partition;
    }

    public void setPartition(Partition partition) {
        this.partition = partition;
    }

    public ParallelLoad getParallel_load() {
        return parallel_load;
    }

    public void setParallel_load(ParallelLoad parallel_load) {
        this.parallel_load = parallel_load;
    }

    public Select getSelect() {
        return select;
    }

    public void setSelect(Select select) {
        this.select = select;
    }

    public void validate(String path){
        path += "source/";
        if(connection_name == null){
            throw new ConfigException("Missing field: " + path + "connection_name");
        }
        if(this.path == null){
            throw new ConfigException("Missing field: " + path + "path");
        }
        if(format == null){
            throw new ConfigException("Missing field: " + path + "format");
        }
        if (!Arrays.asList("json", "parquet", "jdbc", "orc", "libsvm", "csv", "text" , "avro").contains(format)) {
            throw new ConfigException("Wrong format! '" + format + "' is not allowed in " + path + "format");
        }
        if(version == null){
            throw new ConfigException("Missing field: " + path + "version");
        }
        if(partition == null){
            if(select != null){
                select.validateUnpartitioned(path);
            }
        }
        else{
            partition.validate(path);
            if(select == null){
                throw new ConfigException("Missing field: " + path + "select");
            }
            else{
                select.validate(path);
            }
        }
        if(parallel_load != null){
            parallel_load.validate(path);
        }
    }

    @Override
    public String toString() {
        return "Source{" +
                "connection_name='" + connection_name + '\'' +
                ", path='" + path + '\'' +
                ", format='" + format + '\'' +
                ", version='" + version + '\'' +
                ", partition=" + partition +
                ", parallel_load=" + parallel_load +
                ", select=" + select +
                ", autoloading=" + autoloading +
                '}';
    }
}

File: uhrwerk-config/src/main/kotlin/io/qimia/uhrwerk/config/representation/Source.kt
package io.qimia.uhrwerk.config.representation

import com.fasterxml.jackson.annotation.JsonProperty
import io.qimia.uhrwerk.config.builders.ConfigException
import java.util.*

data class Source(
    @JsonProperty("connection_name")
    var connectionName: String? = null,
    var path: String? = null,
    var format: String? = null,
    var version: String? = null,
    var partition: Partition? = null,
    @JsonProperty("parallel_load")
    var parallelLoad: ParallelLoad? = null,
    var select: Select? = null,
    @JsonProperty("auto_load")
    var autoLoad: Boolean = true,
    @JsonProperty("ingestion_mode")
    var ingestionMode: IngestionMode = IngestionMode.ALL,
    var view: String? = null,
    @JsonProperty("fetch_size")
    var fetchSize: Int = 100000,
) {
    fun validate(path: String) {
        var path = path
        path += "source/"
        if (connectionName == null) {
            throw ConfigException("Missing field: " + path + "connection_name")
        }
        if (this.path == null) {
            throw ConfigException("Missing field: " + path + "path")
        }
        if (format == null) {
            throw ConfigException("Missing field: " + path + "format")
        }
        if (!Arrays.asList(
                "json",
                "parquet",
                "jdbc",
                "orc",
                "libsvm",
                "csv",
                "text",
                "avro",
                "redshift"
            )
                .contains(format)
        ) {
            throw ConfigException("Wrong format! '" + format + "' is not allowed in " + path + "format")
        }
        if (version == null) {
            throw ConfigException("Missing field: " + path + "version")
        }
        if (partition == null) {
            if (select != null) {
                select!!.validateUnpartitioned(path)
            }
        } else {
            partition!!.validate(path)
            if (select == null) {
                throw ConfigException("Missing field: " + path + "select")
            } else {
                select!!.validate(path)
            }
        }
        if (parallelLoad != null) {
            parallelLoad!!.validate(path)
        }
    }
}